torch/csrc/jit/codegen/cuda/lower_magic_zero.cpp

```diff
@@ -0,0 +1,116 @@
+#include <torch/csrc/jit/codegen/cuda/lower_magic_zero.h>
+
+#include <torch/csrc/jit/codegen/cuda/dispatch.h>
+#include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_builder.h>
+#include <torch/csrc/jit/codegen/cuda/lower2device.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+class MagicZeroInserter : public kir::MutableIrVisitor {
+ public:
+  static std::vector<kir::Expr*> insert(const std::vector<kir::Expr*>& exprs) {
+    MagicZeroInserter inserter(exprs);
+    return inserter.loop_nests_;
+  }
+
+ private:
+  MagicZeroInserter(const std::vector<kir::Expr*>& exprs)
+      : loop_nests_(exprs), ir_builder(GpuLower::current()->kernel()) {
+    loop_nests_.insert(
+        loop_nests_.begin(), ir_builder.create<kir::InitMagicZero>());
+    for (auto expr : exprs) {
+      handle(expr);
+    }
+  }
+
+  void handle(kir::Expr* expr) {
+    if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+      handle(ite);
+    } else if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+      handle(for_loop);
+    }
+  }
+
+  void handle(kir::IfThenElse* ite) {
+    scope_nest_.push_back(&ite->thenBody());
+    for (auto expr : ite->thenBody().exprs()) {
+      handle(expr);
+    }
+    scope_nest_.pop_back();
+    scope_nest_.push_back(&ite->elseBody());
+    for (auto expr : ite->elseBody().exprs()) {
+      handle(expr);
+    }
+    scope_nest_.pop_back();
+  }
+
+  void handle(kir::ForLoop* fl) {
+    if (fl->start()->isConstScalar() && fl->stop()->isConstScalar() &&
+        !fl->iter_domain()->isThread() &&
+        !(fl->start()->isZeroInt() && fl->stop()->isOneInt())) {
```

Comment:
Perhaps,  a static function could be created to extract common checks from the three locations.